### DISCUSSION TITLE
Avoid redraws when switching tabs on Firefox

### DISCUSSION THREAD

**Discussion by gshpychka** [darkreader/darkreader#8864] [Q&A]
I have a couple dozen open tabs, all with Dynamic mode on. It looks great. But very frequently when I switch between tabs, it loads incompletely (some CSS seems to be missing), and then redraws after about 1 second.
I have plenty of RAM, over 10GB free, so I'm wondering if it is possible to enforce more aggressive caching into RAM (?) to avoid this.

I am using Firefox on Linux. 

Here is an example of what I'm talking about. Excuse the blurriness.

![tab_switching](https://user-images.githubusercontent.com/23005347/167804012-02e43185-3ac8-4c6d-8eb8-45f378379e70.gif)

**Comment by Gusted** [comment]
> I have a couple dozen open tabs, all with Dynamic mode on. It looks great. But very frequently when I switch between tabs, it loads incompletely (some CSS seems to be missing), and then redraws after about 1 second.
> I have plenty of RAM, over 10GB free, so I'm wondering if it is possible to enforce more aggressive caching into RAM (?) to avoid this.

The redrawing is caused by Firefox unloading the tab again, so Dark Reader needs to do it's work all over again. Dark Reader already caches certain expensive tasks, but this is invalidated once the tab is unloaded. Caching "between tabs" isn't possible as browsers tends to kill extensions that are using too much memory.

**Comment by gshpychka** [comment]
> > I have a couple dozen open tabs, all with Dynamic mode on. It looks great. But very frequently when I switch between tabs, it loads incompletely (some CSS seems to be missing), and then redraws after about 1 second.
> 
> > I have plenty of RAM, over 10GB free, so I'm wondering if it is possible to enforce more aggressive caching into RAM (?) to avoid this.
> 
> 
> 
> The redrawing is caused by Firefox unloading the tab again, so Dark Reader needs to do it's work all over again. Dark Reader already caches certain expensive tasks, but this is invalidated once the tab is unloaded. Caching "between tabs" isn't possible as browsers tends to kill extensions that are using too much memory. 

The content isn't unloaded, though, it doesn't refresh the tab, as you can see. It's just some of the DR's CSS, it seems.

That being said, how can I solve this, considering I have plenty of RAM? Maybe some FF config setting?

**Reply by Gusted** [reply]
> The content isn't unloaded, though, it doesn't refresh the tab, as you can see. It's just some of the DR's CSS, it seems.

Well, that's not really something that can happen. Dark Reader doesn't have a unload feature, only the browser can do that by unloading the tab or freeing the tab from memory. 



> That being said, how can I solve this, considering I have plenty of RAM? Maybe some FF config setting?

I have no idea, you would need to check the options in `about:config` for that. Won't surprise me if there's a config option for it.

**Comment by gshpychka** [comment]
> Well, that's not really something that can happen. Dark Reader doesn't have a unload feature, only the browser can do that by unloading the tab or freeing the tab from memory.

I've attached a GIF to the post, you can see that only a part of the CSS is reloaded, not the entire tab. This never happens without Dark Reader.

> I have no idea, you would need to check the options in about:config for that. Won't surprise me if there's a config option for it.

Maybe somebody can chime in and point me to a useful flag. I've looked and tried different flags, to no avail.

**Reply by Gusted** [reply]
> I've attached a GIF to the post, you can see that only a part of the CSS is reloaded, not the entire tab. This never happens without Dark Reader.

IIRC Firefox does _store_ most styles/html of a tab to storage when it unloads a tab, so the GIF isn't far of from what I expect when Firefox does this. But from dark reader's perspective there's nothing to do here, if it's redrawing then it means the tab was unloaded.

**Reply by gshpychka** [reply]
Any idea why this would only happen with Dark Reader enabled?
P.S. Apologies for only now figuring out that I shouldn't create a new answer every time.

**Reply by Gusted** [reply]
The tab unloading? It doesn't only happen with Dark Reader enabled.

**Reply by gshpychka** [reply]
> The tab unloading? It doesn't only happen with Dark Reader enabled.

The behavior in the gif. It does only happen with dark reader enabled, specifically with Dynamic mode.

**Reply by Gusted** [reply]
I'm not sure what you mean by behavior in the GIF, as I [already explained](https://github.com/darkreader/darkreader/discussions/8864#discussioncomment-2733425) that Dark Reader needs to redo all it's work. So you will see Dark Reader's fallback style for a few moments.

**Reply by gshpychka** [reply]
> I'm not sure what you mean by behavior in the GIF, as I [already explained](https://github.com/darkreader/darkreader/discussions/8864#discussioncomment-2733425) that Dark Reader needs to redo all it's work. So you will see Dark Reader's fallback style for a few moments.

It doesn't happen if I disable dark reader, the style is never reloaded like what is shown in the gif.

**Reply by Gusted** [reply]
> It doesn't happen if I disable dark reader, the style is never reloaded like what is shown in the gif.

Well, because the site's style is static. Modern browsers can load styles instantly without the user even noticing.
Dark Reader analyzes these styles and generates a dark mode variant version of it. Which can take some time and thus have this "reloaded" behavior.

**Reply by gshpychka** [reply]
Yup, that's explains why it only happens with the Dynamic mode. Is it possible to recreate the way Dynamic mode looks with hardcoding the styles in Static mode? As in, hardcode the results of Dynamic mode's analysis?

**Reply by gshpychka** [reply]
Regarding tab unloading - I'm still not convinced that's what's happening. The FF documentation says that it only happens when the system memory is low. The behavior in the GIF happens with 10GB+ of free memory. It really does seem like only the custom styling is unloaded.

**Reply by gshpychka** [reply]
I have disabled tab unloading by setting `browser.tabs.unloadOnLowMemory` to `false`, and this still happens. To be clear, the tabs are not reloaded or anything, it's just the Dark Reader's style that's redrawn, and it takes 1-2 seconds.

**Reply by Gusted** [reply]
> I have disabled tab unloading by setting `browser.tabs.unloadOnLowMemory` to `false`, and this still happens. To be clear, the tabs are not reloaded or anything, it's just the Dark Reader's style that's redrawn, and it takes 1-2 seconds.

Something changes 🤷🏽. Dark Reader cannot redrawn itself for no reason.

**Reply by gshpychka** [reply]
Yeah, I'm sure there's a reason. That doesn't help me, though, my goal is to avoid these redraws.
Is it possible to recreate the way Dynamic mode looks with hardcoding the styles in Static mode? As in, hardcode the results of Dynamic mode's analysis?
